web3/_utils/abi.py

```diff
@@ -159,10 +162,69 @@ def collapse_type(base, sub, arrlist):
         return base + str(sub) + ''.join(map(repr, arrlist))
 
 
+def get_tuple_component_types(tuple_type):
+    """Return a list of the types of the components of the given tuple.
+
+    Expects string representation of the tuple, a parenthesized,
+    comma-separated list of the types of the tuples components.  Returns a list
+    representation of those component types.
+
+    >>> get_tuple_component_types('(uint256,uint256)')
+    ['uint256', 'uint256']
+
+    Properly handles the case where some component types are themselves tuples.
+
+    >>> get_tuple_component_types('(uint256,(address,address),uint256)')
+    ['uint256', '(address,address)', 'uint256']
+    """
+    component_types = []
+    components_csv = tuple_type[1:-1]  # strip off expected outer parentheses
+    # consider case of inner parentheses:
+    open_paren_index = components_csv.find("(")
+    close_paren_index = components_csv.rfind(")")
```

Comment:
@feuGeneA  I was looking at this and thought there might be a logic issue here...
```
>>> from web3._utils import abi
>>> tuple = '(int,(address,address),(address,address),int)'
>>> abi.get_tuple_component_types(tuple)
['int', '(address,address),(address,address)', 'int']
```
instead of
```
['int', '(address,address)','(address,address)', 'int']
```
Since this does a reverse lookup of the closing paren, it can't handle a tuple with multiple tuples. This logic should fix that:
```
def get_tuple_component_types(tuple_type):
    components = tuple_type[1:-1]
    
    if '(' not in components and ')' not in components:
        # There were no tuples in the list, so we can just split on commas
        return components.split(',')
    
    split_components = re.split('([\.^(^)])', components)    
    reduced_components = list(filter(lambda val: val not in ['', ',', '(', ')'], split_components))
    clean_components = list(map(lambda val: val.strip(','), reduced_components))
    return list(map(lambda maybeTup: '(' + maybeTup + ')' if ',' in maybeTup else maybeTup, clean_components)) 
```
This works on the following cases:
```
(int,int) # Tuple with no inner tuples
['int', 'int']

(int,(address,address),(address,address),int) # Tuple with consecutive inner tuples
['int', '(address,address)', '(address,address)', 'int']

((address,address),int,(address,address),int,(address,address)) # Tuple with tuples as the first and last types
['(address,address)', 'int', '(address,address)', 'int', '(address,address)']
```
this still falls short in the event that we have inner nested tuples:
```
(int,(address,(int,int)),int) # Tuple with nested inner tuple
['int', 'address', '(int,int)', 'int']
```
and may be inefficient due to the creation of so many lists, but I think it's a good starting point / a step closer.

cc @davesque / @carver

Reply:
@davesque welp... That's a lot simpler than what either of us were doing lol
```
>>> tuple1
'(int,int)'
>>> [str(t) for t in parse(tuple1).components]
['int', 'int']

>>> tuple2
'(int,(address,address),(address,address),int)'
>>> [str(t) for t in parse(tuple2).components]
['int', '(address,address)', '(address,address)', 'int']

>>> tuple3
'((address,address),int,(address,address),int,(address,address))'
>>> [str(t) for t in parse(tuple3).components]
['(address,address)', 'int', '(address,address)', 'int', '(address,address)']

>>> tuple4
'(int,(address,(int,int)),int)'
>>> [str(t) for t in parse(tuple4).components]
['int', '(address,(int,int))', 'int']

>>> tuple5
'(((address,address),int,int),int)'
>>> [str(t) for t in parse(tuple5).components]
['((address,address),int,int)', 'int']
```

I need to do a better job of digging through the helper libraries / repos for web3 to see if there are already functions implemented to do what I'm trying to do :P

Thanks, David!

Reply:
Thanks @davesque !  Much simpler indeed!

@stefanmendoza I just removed the method entirely, since there was only one caller.  And I migrated our test cases over to `tests/core/utilities/test_abi_is_encodable.py`.

Reply:
@feuGeneA @stefanmendoza No problem!  I'm still sort of on holiday but am starting to get back in the groove of things during this week.  I think there might be some other places where the functionality in eth-abi could be used.  Part of my plan for merging this PR is to go through and make sure there aren't still any places where we could be doing that.